src/cargo/util/paths.rs

```diff
@@ -87,6 +89,42 @@ pub fn write(path: &Path, contents: &[u8]) -> CargoResult<()> {
     })
 }
 
+pub fn append(path: &Path, contents: &[u8]) -> CargoResult<()> {
+    (|| -> CargoResult<()> {
+        let mut f = try!(
+                    OpenOptions::new()
+                        .write(true)
+                        .append(true)
+                        .create(true)
+                        .open(path));
+
+        try!(f.write_all(contents));
+        Ok(())
+    }).chain_error(|| {
+        internal(format!("failed to write `{}`", path.display()))
+    })
+}
+
+pub fn write_if_not_exists(path: &Path, contents: &[u8]) -> CargoResult<()> {
+    if file_already_exists(path) {
+        Ok(())
+    } else {
+        write(path, contents)
+    }
+}
+
+pub fn file_already_exists(path: &Path) -> bool {
+    // On error just returns false, expecting subsequent 
+    // creation attempt to fail and deliver proper error message
+    metadata(&path).map(|x| x.is_file()).unwrap_or(false)
+}
+
+pub fn directory_already_exists(path: &Path) -> bool {
```

Comment:
It may be confused with stdlib's `is_dir` then, suprising readers with unusually simple signature.